### ISSUE TITLE
Data loading: numerical vars' variable.rangeMin/Max rounding is excluding data from subsetting display

### ISSUE THREAD

**Issue by bobular** [VEuPathDB/EdaDataService#76]
How to reproduce the bug:

1. GEMS1
2. subsetting, go to **Household wealth, numerical**
3. set the binWidth to 5 (the binWidth does not matter but a large one makes the sums easier)
3. with mouse, select whole of the histogram
4. Note that we are expecting to retain 22,560 households by selecting on the whole histogram
5. But actually only 22,539 HH are in the subset (entity diagram)
6. Mouse-over the bars to add up the bin-sizes: 21767+768+4 = 22539

The issue is that the first bin starts at -2.997481 - which ultimately comes from `variable.rangeMin` (sent to `distribution` endpoint as `binSpec.displayRangeMin`).  This number has been rounded to a value that is greater than the value of this variable for 21 data points.  Thus the left-most edge of the first bin is to the right of some of the data.

You can work around this by manually extending the x-axis to the left (using the controls) - then if you select the whole histogram you get the whole 22,560 households in the selection, and the bar heights add up to the same number too. 

The same may also be true for the right-hand end of scale - but I haven't seen an example yet.

It's my opinion that the `rangeMin` for this variable should have been rounded to -2.997482

**Comment by d-callan** [comment]
looks like the min value is coming from r.. and r does like to automatically round and format things. ex:
> \> min(1.233573262400573663)
[1] 1.233573

HOWEVER!! that is automatically calling `print`, which is what is doing the formatting/rounding. its internally stored as the actual value. i see in [here](https://github.com/VEuPathDB/ApiCommonData/blob/04aaa7ea104bb3f1af5d16febae698a994199f3b/Load/plugin/perl/LoadAttributesFromEntityGraph.pm) that we use base r's `write` function to write out these values. that will unfortunately also automatically format/round the values. **i recommend instead using `data.table::fwrite` which will write the actual stored/ unrounded value.**

**Comment by bobular** [comment]
Hi @steve-fischer-200 @d-callan 

Either the variable annotation should contain the full precision min and max, or somehow the rounding to fixed or significant figures should be done in the spirit of `floor` and `ceiling` respectively.   

Here are some examples using 3 significant figures
| raw number  | rounded for min | rounded for max |
| --- | --- | --- |
| 1.44444444 | 1.44 | 1.45 |
| 1.55555555 | 1.55 | 1.56 |
| -1.44444444 | -1.45 | -1.44 |
| -1.55555555 | -1.56 | -1.55 |

Though I note that there's no actual R function to do this.
Some discussion here: https://stackoverflow.com/questions/39802236/round-down-a-numeric

**Comment by bobular** [comment]
Note that we might be able to do something in the client (for Phase 1) if we know the numbers in `variable.rangeMin/Max` have N significant (or fixed) digits, we can re-round to fewer than N digits.

**Comment by steve-fischer-200** [comment]
i just want to clarify that the adjustments being discussed here are in the data loading phase, when rangeMin and rangeMax are computed and stored in the db

**Comment by jaycolin** [comment]
This does not appear to be a loading issue. I have compared every min, max from another study where I already have summary stats before loading (GEMS1A), compared to loaded values I did not find any discrepancies in floating point values. (There are some differences where raw min=0 but loaded min=1 which seem to be from sets of integer values e.g. "Water trips per week (EUPATH_0010137)", and some dates which are correctly obfuscated; and cases where raw min=1.00 but the loaded value is just "1")
